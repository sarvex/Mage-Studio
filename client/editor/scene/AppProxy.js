const config = {
    screen: {
        frameRate : 60,
        alpha: true
    },

    lights: {
        shadows: true
    },

    physics: {
        enabled: false
    },

    tween: {
        enabled: false
    },

    camera : {
        //handling useful informations about our camera.
        fov : 75,
        near : 0.1,
        far : 3000000
    },

    ui: {
        enabled: false
    },

    selector: '#gameContainer'
};

const assets = {
    models: {
        'lightHolder': '/models/lamp.gltf'
    },
    textures: {
        'cube_prototype_green': '/textures/cube_prototype_green.png',
        'cube_prototype_dark': '/textures/cube_prototype_dark.png',
        'cube_prototype_light': '/textures/cube_prototype_light.png',
        'cube_prototype_orange': '/textures/cube_prototype_orange.png',
        'cube_prototype_purple': '/textures/cube_prototype_purple.png',
    }
};

let app;

export async function getOrCreateApp() {

    if (window && !app) {
        return Promise.all([
            import('mage-engine'),
            import('./EditorScene/App')
        ]).then(([{ Router }, { EditorScene }]) => {
            Router.on('/', EditorScene);
            console.log('creating app');
            return Router
                .start(config, assets)
                .then((application) => {
                    app = application;
                    console.log(application);
                    return Promise.resolve(app);
                })
        });
    }

    // console.log('we have the app');

    return Promise.resolve(app);
}
