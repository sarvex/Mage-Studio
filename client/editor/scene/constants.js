export const EVENTS = {
    HIERARCHY_CHANGE: "HIERARCHY_CHANGE",
    ELEMENT: {
        SELECTED: "ELEMENT.SELECTED",
    },
};
