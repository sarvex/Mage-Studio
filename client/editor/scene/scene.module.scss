@import '../../shared/constants.scss';
@import '../../shared/theme.scss';
@import '../panels/panel.module.scss';

.scene-container {

    flex: 3 1 auto;
    border: 0;
    margin-bottom: 0;
    background-color: $scene-container-background;
    max-width: 73vw;
    max-height: calc(73vh - 64px);

    // &.fullscreen {
    //     position: absolute;
    //     left: 0;
    //     z-index: 100;
    //     & .gameContainer,
    //     & .gamePlayer {
    //         height: 100%;
    //         width: 100%;
    //     }
    // }

    & .gameContainer {
        flex: 1 1 auto;
        max-width: 73vw;
        max-height: calc(73vh - 64px);
        margin-bottom: 2px;
    }

    // & .gamePlayer+.gameContainer {
    //     visibility: hidden;
    // }

    // & .gamePlayer {
    //     position: relative;
    //     margin: auto;
    //     height: 95.6vh;
    //     width: 66.6vw;
    // }
    // & .scene-toolbar {
    //     /* margin: auto; */
    //     /* display: inline-block; */
    //     position: absolute;
    //     top: 0;
    //     margin-left: $base-unit / 2;
    //     .scene-toolbar-action {
    //         /* width: 24px; */
    //         padding: $base-unit / 2;
    //         min-width: $base-unit * 4;
    //         //height: $base-unit * 4;
    //         //line-height: $base-unit * 2;
    //         text-align: center;
    //         margin-top: $base-unit / 2;
    //         color: $white;
    //         font-size: 10px;
    //         background: $deep-blue-13;
    //         float: left;
    //         border-right: 1px solid $deep-blue-10;
    //         &:last-child {
    //             border: none;
    //         }
    //         &.active {
    //             border-bottom: 1px solid $deep-blue-5;
    //             background: $deep-blue-15;
    //         }
    //         &:hover {
    //             background: $deep-blue-15;
    //             cursor: pointer;
    //         }
    //     }
    // }
}