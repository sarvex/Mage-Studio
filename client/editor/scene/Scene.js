import React from 'react';
import debounce from '../../lib/debounce';
import { getOrCreateApp } from './AppProxy';

import style from './scene.module.scss';

export class Scene extends React.Component {

    constructor(props) {
        super(props);

        this.app = {};
    }

    componentDidMount() {
        const { store, onMeshChanged, onElementAttached, onElementDetached, onSceneExported, config, onSceneLoad } = this.props;

        getOrCreateApp()
            .then(app => {
                this.app = app;
                // this.app.addEventListener('elementChanged', debounce(onMeshChanged, 15));
                this.app.addEventListener('elementAttached', onElementAttached);
                this.app.addEventListener('elementDetached', onElementDetached);
            });
    }

    async componentDidUpdate(prevProps) {
        if (prevProps.fullscreen !== this.props.fullscreen && this.app) {
            this.app.onResize();
        }
    }

    render() {
        return <div
            id="gameContainer"
            className={style.gameContainer}
            tabIndex={0}></div>
    }
}

export default Scene;
