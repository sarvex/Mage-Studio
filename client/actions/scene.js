import {
    SCENE_SAVE_REQUEST,
    SCENE_SAVE_LOADING,
    SCENE_SAVE_SUCCESS,
    SCENE_SAVE_FAILURE,
    PROJECT_RUNNING,
    PROJECT_STOPPED,
    PROJECT_PLAYER_VISIBLE,
    PROJECT_PLAYER_HIDDEN,
} from "./types";
import { getProjectsUrl, getScenesUrl, MIMETYPES } from "../lib/constants";
import { getOrCreateApp } from "../editor/scene/AppProxy";
import axios from "axios";
import { ELEMENTS } from "../contants";

export const requestSceneJson = () => ({
    type: SCENE_SAVE_REQUEST,
});

export const sceneSaveLoading = () => ({
    type: SCENE_SAVE_LOADING,
});

export const sceneSaveSuccess = () => ({
    type: SCENE_SAVE_SUCCESS,
});

export const sceneSaveFailure = () => ({
    type: SCENE_SAVE_FAILURE,
});

export const projectRunning = url => ({
    type: PROJECT_RUNNING,
    url,
});

export const projectStopped = () => ({
    type: PROJECT_STOPPED,
});

export const projectPlayerVisible = visible => dispatch => {
    dispatch({
        type: visible ? PROJECT_PLAYER_VISIBLE : PROJECT_PLAYER_HIDDEN,
    });
};

export const addElement = type => {
    getOrCreateApp().then(app => {
        switch (type) {
            case ELEMENTS.BASE.CUBE:
                app.addCube();
                break;
            case ELEMENTS.BASE.SPHERE:
                app.addSphere();
                break;
            case ELEMENTS.BASE.CYLINDER:
                app.addCylinder();
                break;
            case ELEMENTS.BASE.CONE:
                app.addCone();
                break;
            case ELEMENTS.BASE.BOX:
                app.addBox();
                break;
            case ELEMENTS.BASE.PLANE:
                app.addPlane();
                break;
            case ELEMENTS.LIGHTS.AMBIENT:
                app.addAmbientLight();
                break;
            case ELEMENTS.LIGHTS.SUN:
                app.addSunLight();
                break;
            default:
                break;
        }
    });
};

export const saveScene = (name, scene) => dispatch => {
    dispatch(sceneSaveLoading());

    const blobParts = [JSON.stringify({ ...scene })];
    const blobOptions = { type: MIMETYPES.APPLICATION_JSON };
    const blob = new Blob(blobParts, blobOptions);
    const formData = new FormData();

    formData.append("data", blob, name + ".json");

    axios
        .post(getScenesUrl(name), formData)
        .then(() => dispatch(sceneSaveSuccess()))
        .catch(() => dispatch(sceneSaveFailure()));
};

export const loadScene = name => () => {
    const url = getScenesUrl(name);

    const sceneOptions = {
        scriptEnabled: false,
        useHelper: true,
    };
    // should dispatch an action here and show some loading screen when
    // loading the scene from BE.
    getOrCreateApp().then(app => {
        axios.get(url).then(({ data }) => app.parseScene(data, sceneOptions));
    });
};

export const startProject = project => dispatch => {
    const url = `${getProjectsUrl(project)}/start`;

    // dispatch something here about starting the app

    axios
        .post(url)
        .then(({ data }) => {
            // dispatch something that will cause the editor to show the iframe
            dispatch(projectRunning(data.url));
            dispatch(projectPlayerVisible(true));
        })
        .catch(e => {
            // handling the failure of starting the project
        });
};

export const stopProject = project => dispatch => {
    const url = `${getProjectsUrl(project)}/stop`;

    // dispatch something here about starting the app
    axios
        .post(url)
        .then(({ data }) => {
            // dispatch something that will cause the editor to show the iframe
            dispatch(projectStopped());
            dispatch(projectPlayerVisible(false));
        })
        .catch(e => {
            // handling the failure of starting the project
        });
};
