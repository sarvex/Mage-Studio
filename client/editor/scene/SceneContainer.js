import React from 'react';
import {
    connect
} from 'react-redux';

import classnames from 'classnames';

// import {
//     meshChanged,
//     meshAttached,
//     meshDetached
// } from '../../actions/currentMesh';

import { elementAttached, toggleInspectorVisibility } from '../../actions/inspector';

import {
    saveScene,
    loadScene,
    startProject,
    stopProject
} from '../../actions/scene';

import Scene from './Scene';
// import Player from './Player';

import style from './scene.module.scss';
// import { showModelUploadModal } from '../../actions/models';

export class SceneContainer extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            fullscreen: false
        };
    }

    handleFullScreen = (fullscreen) => {
        this.setState({ fullscreen });
    }

    render() {
        const {
            scene,
            store,
            config,
            loadScene,
            onElementChanged,
            onElementAttached,
            onElementDetached,
            onSceneExported,
            startProject,
            showModelModal,
            stopProject
        } = this.props;

        const { fullscreen } = this.state;
        // const { projectPlayerVisible, projectUrl } = scene;

        const className = classnames(style['scene-container'], {
            [style.fullscreen]: fullscreen
        });

        return (
            <div className={className}>
                {/* <Player
                    url={projectUrl}
                    visible={projectPlayerVisible} /> */}
                <Scene
                    fullscreen={fullscreen}
                    config={config}
                    store={store}
                    onSceneLoad={loadScene}
                    onElementChanged={onElementChanged}
                    onElementAttached={onElementAttached}
                    onElementDetached={onElementDetached}
                    onSceneExported={onSceneExported} />
            </div>
        );
    }
}

const mapStateToProps = ({ config, scene }) => ({
    // config,
    // scene
});

const mapDispatchToProps = (dispatch) => ({
    // startProject: project => dispatch(startProject(project)),
    // showModelModal: () => dispatch(showModelUploadModal()),
    // onElementChanged: ({name, position, rotation, scale}) => {
    //     return dispatch(meshChanged(name, position, rotation, scale))
    // },
    onElementAttached: () => dispatch(elementAttached()),
    onElementDetached: () => dispatch(toggleInspectorVisibility(false)),
    // onSceneExported: (name) => ({ data }) => dispatch(saveScene(name, data)),
    // loadScene: (name) => dispatch(loadScene(name)),
    // stopProject: (project) => dispatch(stopProject(project))
});

export default connect(mapStateToProps, mapDispatchToProps)(SceneContainer);
