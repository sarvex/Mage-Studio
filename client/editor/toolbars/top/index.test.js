import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import sinon from "sinon";

import { addElement } from "../../scene/AppProxy";
import { Toolbar } from "./index";

jest.mock("../../scene/AppProxy");

describe("TopToolbar", () => {
    beforeEach(() => {
        addElement.mockClear();
    });

    describe("render", () => {
        it("should render properly", () => {
            const component = shallow(<Toolbar />);
            expect(toJSON(component)).toMatchSnapshot();
        });
    });

    describe("AddElementDropdown", () => {
        it("should call addElement when onElementSelection occurs", () => {
            const component = shallow(<Toolbar />);

            component.find("AddElementDropdown").simulate("elementSelection");

            expect(addElement).toHaveBeenCalledTimes(1);
        });
    });

    // describe('SpaceSettingsButton', () => {

    //     it('should call onTransformSpaceChange action when user changes transform space', () => {
    //         const spy = sinon.spy();
    //         const component = shallow(<Toolbar onTransformSpaceChange={spy} />)

    //         component.find('SpaceSettingsButton').simulate('transformSpaceChange');

    //         expect(spy.called).toBe(true);
    //     });
    // });

    // describe('ControlsRadioGroup', () => {

    //     it('should call onTransformControlChange action when user changes transform control', () => {
    //         const spy = sinon.spy();
    //         const component = shallow(<Toolbar onTransformControlChange={spy} />)
    //         component.find('ControlsRadioGroup').simulate('transformControlChange', { target: { value: '' } });

    //         expect(spy.called).toBe(true);
    //     });
    // });

    // describe('SnapControlsGroup', () => {

    //     it('should call onTransformSnapChange action when user decreases snap value', () => {
    //         const spy = sinon.spy();
    //         const component = shallow(<Toolbar onTransformSnapChange={spy} />)

    //         component.find('SnapControlsGroup').simulate('snapDecrease');

    //         expect(spy.called).toBe(true);
    //     });

    //     it('should call onTransformSnapChange action when user increases snap value', () => {
    //         const spy = sinon.spy();
    //         const component = shallow(<Toolbar onTransformSnapChange={spy} />)

    //         component.find('SnapControlsGroup').simulate('snapIncrease');

    //         expect(spy.called).toBe(true);
    //     });

    //     it('should call onTransformSnapChange action when user toggles snap', () => {
    //         const spy = sinon.spy();
    //         const component = shallow(<Toolbar onTransformSnapChange={spy} />)

    //         component.find('SnapControlsGroup').simulate('snapEnabledChange');

    //         expect(spy.called).toBe(true);
    //     });
    // });
});
