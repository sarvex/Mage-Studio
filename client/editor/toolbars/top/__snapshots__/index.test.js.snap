// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopToolbar render should render properly 1`] = `
<div
  className="toolbar"
>
  <ul
    className="settings-list"
  >
    <li
      className="settings-list-item"
    >
      <SpaceSettingsButton
        onTransformSpaceChange={[Function]}
      />
    </li>
    <li
      className="settings-list-item"
    >
      <AddElementDropdown
        onElementSelection={[MockFunction]}
      />
    </li>
    <li
      className="settings-list-item"
    >
      <ControlsRadioGroup
        onTransformControlChange={[Function]}
      />
    </li>
    <li
      className="settings-list-item"
    >
      <SnapControlsGroup
        onSnapDecrease={[Function]}
        onSnapEnabledChange={[Function]}
        onSnapIncrease={[Function]}
      />
    </li>
  </ul>
</div>
`;
